vendor/github.com/docker/libnetwork/resolver.go

```diff
@@ -491,9 +491,9 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
 			}
 			r.forwardQueryEnd()
 			if resp != nil {
-				if resp.Rcode == dns.RcodeServerFailure {
-					// for Server Failure response, continue to the next external DNS server
-					logrus.Debugf("[resolver] external DNS %s:%s responded with ServFail for %q", proto, extDNS.IPStr, name)
+				if resp.Rcode == dns.RcodeServerFailure || resp.Rcode == dns.RcodeNameError || resp.Rcode == dns.RcodeRefused {
```

Comment:
These changes are in vendored files, so will have to be merged in libnetwork first, and then pulled in.

Note that `NXDOMAIN` should result in a failure: NXDOMAIN is an authoritive answer, so when receiving an NXDOMAIN, we're done.

- From RFC 1035: Name Error - Meaningful only for responses from an authoritative name server, this code signifies that the domain name referenced in the query does not exist.
- FROM RFC 8020: When an iterative caching DNS resolver receives an NXDOMAIN response, it SHOULD store it in its cache and then all names and resource record sets (RRsets) at or below that node SHOULD be considered unreachable. Subsequent queries for such names SHOULD elicit an NXDOMAIN response.

I opened a PR with some changes in https://github.com/docker/libnetwork/pull/2171 (so that it can be discussed)